core/src/main/scala/org/apache/spark/deploy/Client.scala

```diff
@@ -80,11 +81,30 @@ private class ClientEndpoint(
         val extraJavaOptsConf = "spark.driver.extraJavaOptions"
         val extraJavaOpts = sys.props.get(extraJavaOptsConf)
           .map(Utils.splitCommandString).getOrElse(Seq.empty)
-        val sparkJavaOpts = Utils.sparkJavaOpts(conf)
+
+        // we don't want to send any secrets as system properties because they would be translated
+        // to Java options which are passed in command-line and thus visible for other users
+        val propsToExclude = Set(Namespace.Blank, Namespace.Submission, Namespace.Application)
+          .map(_.inConf(SPARK_AUTH_SECRET_CONF))
+        val sparkJavaOpts = Utils.sparkJavaOpts(conf, key => !propsToExclude.contains(key))
         val javaOpts = sparkJavaOpts ++ extraJavaOpts
-        val command = new Command(mainClass,
-          Seq("{{WORKER_URL}}", "{{USER_JAR}}", driverArgs.mainClass) ++ driverArgs.driverOptions,
-          sys.env, classPathEntries, libraryPathEntries, javaOpts)
+
+        // we want to transport secrets only from submission and app namespaces; they will be
+        // transported as environment variables though
+        val envSecrets = Set(Namespace.Submission, Namespace.Application).flatMap { ns =>
```

Comment:
I think used `Seq` in these places would be better, since it's faster and you don't really need the functionality of a `Set` here.